### PR TITLE
 Harnesses for carrying_mul

### PR DESCRIPTION
**Pull request by lanfeima**
### Parent branch: main

---

### Changes

- Added macros for generating `carrying_mul` harnesses
- Added harnesses for `carrying_mul` for the following integer types:
  - `u8`, `u16`, `u32`, `u64`

### Revalidation

Per the discussion in issue [#59](https://github.com/model-checking/verify-rust-std/issues/59), we have to build and run Kani from the `feature/verify-rust-std` branch.

To revalidate the verification results, run the following command. `<harness_to_run>` can be either `num::verify` to run all harnesses or `num::verify::<harness_name>` (e.g., `carrying_mul_u8_small`) to run a specific harness.

```bash
kani verify-std "path/to/library" \
    --harness <harness_to_run> \
    -Z unstable-options \
    -Z function-contracts \
    -Z mem-predicates
```

### CODE REVIEW

File: library/core/src/num/mod.rs [hunk 1]
```diff
@@ -1658,6 +1658,40 @@ mod verify {
         }
     }
 
+    macro_rules! generate_carrying_mul_intervals {
```
Comment (celinval) [inline library/core/src/num/mod.rs:1661]: can you please document this macro?

File: library/core/src/num/mod.rs [hunk 2]
```diff
@@ -1849,4 +1883,33 @@ mod verify {
     generate_unchecked_math_harness!(u64, unchecked_sub, checked_unchecked_sub_u64);
     generate_unchecked_math_harness!(u128, unchecked_sub, checked_unchecked_sub_u128);
     generate_unchecked_math_harness!(usize, unchecked_sub, checked_unchecked_sub_usize);
+
+
+    // ====================== u8 Harnesses ======================
+    generate_carrying_mul_intervals!(u8, u16,
+        carrying_mul_u8_small, 0u8, 10u8,
+        carrying_mul_u8_large, u8::MAX - 10u8, u8::MAX
```
Comment (celinval) [inline library/core/src/num/mod.rs:1891]: How long would it take for a harness that does not restrict the interval for the `u8` case?

**Review by celinval — COMMENTED** [review]
I was wondering if we can use wider intervals.

File: library/core/src/num/mod.rs (hunk 1 above)
Comment (lanfeima) [reply library/core/src/num/mod.rs:1661]: Thank you for the feedback! I've added documentation to the macro generate_carrying_mul_intervals.

File: library/core/src/num/mod.rs (hunk 2 above)
Comment (lanfeima) [reply library/core/src/num/mod.rs:1903]: I modified the code to include a new harness carrying_mul that tests carrying_mul for the full range of u8 values (from 0u8 to u8::MAX).
It hang for around 1 min, and 
Here are the results of the verification run using Kani:

Harness name: carrying_mul_u8_full_range
Verification Time: ~2.8- 2.95 seconds
Result: All checks passed (0 of 45 failed)

Comment (lanfeima) [reply library/core/src/num/mod.rs:1903]: Would it be a good idea to widen the range as possible, as long as the execution time is reasonable?

Comment (carolynzech) [reply library/core/src/num/mod.rs:1903]: Yes, generally speaking we always want to verify a function against its full range of inputs--you should only limit the range if the verification time is prohibitively high. ~1 min is fast enough that I would say to go for the full range.

Comment (lanfeima) [reply library/core/src/num/mod.rs]: Thank you for the feedback! I have made the following changes to the harnesses: For u8 and u16, I have verified the carrying_mul function using the full range of values, as the verification time for these types was reasonable. For u32 and u64, the verification time for the full range was prohibitively high(u32 stuck at "Running propositional reducation" for 4 hours without progress ),I have implemented smaller intervals to ensure the verification can complete within a reasonable time.

**Review by carolynzech — APPROVED** [review]

**Review by celinval — APPROVED** [review]
Thank you!

**Pull request merged**